File: course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] result = new int[numCourses];
        int[] indegrees = new int[numCourses];
        List<List<Integer>> adj = new ArrayList<>();
        int nodesVisited = 0;

        //build adj list, indegree list
        for (int i = 0; i < numCourses; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] prereq : prerequisites) {
            adj.get(prereq[1]).add(prereq[0]);
            indegrees[prereq[0]]++;
        }

        Queue<Integer> queue = new ArrayDeque<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegrees[i] == 0) {
                queue.offer(i);
            }
        }

        while (!queue.isEmpty()) {
            int node = queue.poll();
            result[nodesVisited] = node;
            nodesVisited++;
            for (int neighbor : adj.get(node)) {
                indegrees[neighbor]--;
                if (indegrees[neighbor] == 0) {
                    queue.offer(neighbor);
                }
            }
        }
        if (nodesVisited == numCourses) {
            return result;
        }

        return new int[] {};
    }
}

File: 210-course-schedule-ii/course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = defaultdict(list)
        indegree = {}
        q = deque()
        result = []

        for dest, source in prerequisites:
            adj[source].append(dest)
            if dest not in indegree:
                indegree[dest] = 0
            indegree[dest] += 1

        for node in range(numCourses):
            if node not in indegree or indegree[node] == 0:
                q.append(node)

        while q:
            curr = q.popleft()
            result.append(curr)
            for neighbor in adj[curr]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    q.append(neighbor)
        return result if len(result) == numCourses else []